src/test/java/com/sivalabs/bookstore/catalog/domain/ProductServiceTest.java

```diff
@@ -0,0 +1,79 @@
+package com.sivalabs.bookstore.catalog.domain;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import com.sivalabs.bookstore.catalog.common.TestHelper;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.mongodb.core.query.TextCriteria;
+
+public class ProductServiceTest {
+
+    private ProductService productService;
+    private ProductRepository productRepository;
+    private ProductMapper productMapper;
+
+    @BeforeEach
+    public void setup() {
```

Comment:
Initializing for each Junit Testcase is not recommended as it will not change throughout all TestCases. Preferably use @Mock and @InjectMock annotations.

Reply:
JUnit 5 by default create separate instances per-test method unless you add @TestInstance(Lifecycle.PER_CLASS). See https://stackoverflow.com/questions/52551718/what-use-is-testinstance-annotation-in-junit-5

Reply:
Hi @rajadilipkolli , thank you for the suggestion. I am curious as to what are the benefits of using `@Mock and @InjectMock` ? Are there any best practices around this ? 

Also, I tried the annotation, is this what you meant ?
```java
    @InjectMocks  private ProductService productService;
    
    @Mock  private ProductRepository productRepository;
    @Mock  private ProductMapper productMapper;
    
    @BeforeEach
    public void setup() {
        MockitoAnnotations.openMocks(this);
        // this.productMapper = mock(ProductMapper.class);
        // this.productRepository = mock(ProductRepository.class);
        // this.productService = new ProductService(productRepository, productMapper);
    }
```

Also, When we do MockitoAnnotations.openMocks(this), Interesting thing is that this also seems to be creating new instance of Mocked objects. I might be totally off track here though.

Reply:
Thanks for sharing link @sivaprasadreddy

Reply:
When we use @Mock and @InjectMocks we don't need `MockitoAnnotations.openMocks(this)`, so whole @BeforeEach setup method can be removed.

This approach of using annotations makes code little cleaner and easy to extend in future. example if you have added one more dependency then adding one line like `@Mock MyNewService newService` is enough.

Reply:
@rajadilipkolli  If I don't add `MockitoAnnotations.openMocks(this)` , in that case then mocked objects throw a NullPointer. Googling suggests annotating class with `@RunWith(MockitoJUnitRunner.class)` is another option to resolve this issue.

I think I just prefer the clear readability/verbosity of plain old constructor and keeps it easier to maintain the code over a very long time. 

Having said that, I am not tied to that approach, so very happy to incorporate it. Thanks for this discussion 👍